core/src/main/java/com/linecorp/armeria/client/AllInOneClientFactory.java

```diff
@@ -84,38 +84,91 @@ public AllInOneClientFactory(SessionOptions options) {
      * @param useDaemonThreads whether to create I/O event loop threads as daemon threads
      */
     public AllInOneClientFactory(SessionOptions options, boolean useDaemonThreads) {
+        super(options, useDaemonThreads);
+
         // TODO(trustin): Allow specifying different options for different session protocols.
         //                We have only one session protocol at the moment, so this is OK so far.
-        final HttpClientFactory httpClientFactory = new HttpClientFactory(options, useDaemonThreads);
-        final THttpClientFactory thriftClientFactory = new THttpClientFactory(httpClientFactory);
+
+        // Do not let the delegates create or manage event loops by specifying the EventLoop explicitly.
+        // See NonDecoratingClientFactory.closeEventLoopGroup for more information.
+        final SessionOptions delegateOptions =
+                SessionOptions.of(options, SessionOption.EVENT_LOOP_GROUP.newValue(eventLoopGroup()));
+
+        final List<ClientFactoryProvider> providers =
+                Streams.stream(ServiceLoader.load(ClientFactoryProvider.class,
+                                                  AllInOneClientFactory.class.getClassLoader()))
+                       .collect(Collectors.toCollection(ArrayList::new));
+
+        if (providers.isEmpty()) {
+            throw new IllegalStateException("could not find any providers");
+        }
+
+        final List<ClientFactory> factories = new ArrayList<>();
+        int numFactories = 0;
+        do {
+            for (Iterator<ClientFactoryProvider> i = providers.iterator(); i.hasNext();) {
+                final ClientFactoryProvider p = i.next();
+                final Map<Class<? extends ClientFactory>, ClientFactory> dependencies =
+                        buildDependencies(p, factories);
+                if (dependencies == null) {
```

Comment:
Will use `requiredClientFactories` and `findRequiredClientFactories()`. Sounds good?